library/ccm.c

```diff
@@ -403,7 +406,13 @@ static const unsigned char res[NB_TESTS][32] = {
 int mbedtls_ccm_self_test( int verbose )
 {
     mbedtls_ccm_context ctx;
-    unsigned char out[32];
+    /*
+     * Some HW accelerators require the input and output buffers to be aligned
```

Comment:
Are we sure we want to accommodate HW acceleration drivers so much? Our software implementation doesn't have this requirement. What if some HW accelerator required buffers to be aligned in such a way that doesn't satisfy whichever HW we are now catering to? Now we'd have to support both contradictory HW accelerator requirements, which is impossible.

Alternatives to consider (we could do one or more of these):
- ALT implementations provide their own self tests
- ALT implementations return an error when given data they can't process (like unaligned)
- ALT implementations copy data to buffers that the HW accelerator is capable of operating on. (This is common with hardware that uses DMA: a DMA-capable buffer must be used, and the user buffer copied into the DMA-capable buffer, since the user buffer isn't suitable for use with DMA.)

Reply:
I think this is a good point. I think architecturally alternative implementations should be able to implement their own self-test without implementing the whole file, and also be able to return an error. 

This is the wrong point in the development cycle to make this change though, so I'd suggest accepting the PR, and raising this as a defect.

Reply:
I don't think the ALT implementations should provide their own self tests. I think tests should be defined by different party than the ALT developer. I think it's reasonable that Mbed TLS \ Mbed OS will test the alternative cryptography

The ALT implementation may know if the input is unaligned, but it may not know if it's not contiguous, for example. I don't think it's reasonable to have a special code.
Having the ALT implementation copy the data to their own local buffer will increase internal stack size and add limitation on the length of the input buffer. For the bare metal devices, the buffers are already DMA-capable. For high level driver running on Cortex-A, I agree that it is in the ALT implementation to handle it.

I agree we shouldn't accommodate every HW accelerator, but common use cases ( e.g. contiguous buffers) should be supported

Reply:
> I don't think the ALT implementations should provide their own self tests. I think tests should be defined by different party than the ALT developer.

The self-test isn't an acceptance test. It's a useful time to quickly check at run-time if the system is working as intended. So, the importance of avoiding a "conflict of interest" between tester and implementer isn't as important here. On the contrary, the implementer is going to have intimate knowledge of their hardware and can kick off a hardware-specific self-test should we provide them the hooks to do so. The implementer will also know if there are special constraints on the input data and can format their input data correctly for use with their hardware.

> Having the ALT implementation copy the data to their own local buffer will increase internal stack size

Stack isn't the only choice. Static buffers can be used, or heap if available.

> ...common use cases ( e.g. contiguous buffers) should be supported

I don't think we have a good feel for how common that is. Maybe CC 310 is the only one with this particular constraint. Maybe there are other issues with buffers when feeding CC 310 from the TLS stack, not just the self-test. In such a case, it'd be better for the driver to smooth over any differences. It would be silly to force our TLS stack to always use a specific alignment for our buffers; much better would be a compile-time configurable alignment constraint that applies not just to the self-test, but everywhere in the library.

Reply:
> I don't think the ALT implementations should provide their own self tests. I think tests should be defined by different party than the ALT developer. I think it's reasonable that Mbed TLS \ Mbed OS will test the alternative cryptography

I agree with Jaeden. What are we testing here? The self-test is intended to be used by application software to ascertain that the feature is working correctly at run-time. It's not about demonstrating the design or implementation are correct, but that the software is operating correctly. As such a self-test might be run on embedded systems as a diagnostic feature or at boot-up.

For a cryptographic accelerator these tests might not make sense, or you might want additional tests. For instance - maybe you'd want to access a self-test on the accelerator. Apart from this function, how else would you do it?

Reply:
@Patater - No - self-tests need to be controllable by application software. On many devices there will be a 'service mode' or 'diagnostic mode', where you can run hardware tests, sometimes in addition to start-up tests. We need to provide a function, in my view, outside the platform initialisation.

Reply:
I agree with @Patater's view. I think its not feasible for us to ensure that the self tests run on every possible/supported hardware. It should be the person who maintains that system or does the port the one that provides suitable self tests for their platform. However, I also think that there is very little support from the Mbed TLS side to do so. Ideally there would be hooks or something to facilitate that task (again, as mentioned by @Patater)